### PR TITLE
Add generic support to `@method` definitions

### PR DESCRIPTION
**Pull request by tyabu12**
This allows implementing https://github.com/phpstan/phpstan/issues/6371 to add support`@method` generics.

### CODE REVIEW

File: src/Ast/PhpDoc/MethodTagValueGenericNode.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\PhpDocParser\Ast\PhpDoc;
+
+use PHPStan\PhpDocParser\Ast\Node;
+use PHPStan\PhpDocParser\Ast\NodeAttributes;
+use PHPStan\PhpDocParser\Ast\Type\TypeNode;
+
+class MethodTagValueGenericNode implements Node
```
Comment (ondrejmirtes) [inline src/Ast/PhpDoc/MethodTagValueGenericNode.php]: I feel like this doesn't need a new node, but TemplateTagValueNode can be reused instead. We'd also gain default template types (https://github.com/phpstan/phpdoc-parser/pull/148) automatically :)

File: src/Ast/PhpDoc/MethodTagValueNode.php [hunk 2]
```diff
@@ -20,17 +20,21 @@ class MethodTagValueNode implements PhpDocTagValueNode
 	/** @var string */
 	public $methodName;
 
+	/** @var MethodTagValueGenericNode[] */
+	public $generics;
```
Comment (ondrejmirtes) [inline src/Ast/PhpDoc/MethodTagValueNode.php]: This thing should be called `$templateTypes` probably.

File: src/Ast/PhpDoc/MethodTagValueNode.php [hunk 3]
```diff
@@ -20,17 +20,21 @@ class MethodTagValueNode implements PhpDocTagValueNode
...
+
 	/** @var MethodTagValueParameterNode[] */
 	public $parameters;
 
 	/** @var string (may be empty) */
 	public $description;
 
-	public function __construct(bool $isStatic, ?TypeNode $returnType, string $methodName, array $parameters, string $description)
+	public function __construct(bool $isStatic, ?TypeNode $returnType, string $methodName, array $generics, array $parameters, string $description)
```
Comment (ondrejmirtes) [inline src/Ast/PhpDoc/MethodTagValueNode.php]: This parameter needs to be added as last and needs to be optional, otherwise it's a BC break.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
And please fix the build failures :)

**Comment by tyabu12** [comment]
@ondrejmirtes 
Thank you for quick reviewing!
I fixed your review comments and build failures.

File: src/Parser/PhpDocParser.php [hunk 4]
```diff
@@ -346,6 +346,14 @@ private function parseMethodTagValue(TokenIterator $tokens): Ast\PhpDoc\MethodTa
 			exit;
 		}
 
+		$templateTypes = [];
+		if ($tokens->tryConsumeTokenType(Lexer::TOKEN_OPEN_ANGLE_BRACKET)) {
+			do {
+				$templateTypes[] = $this->parseMethodTagValueTemplateType($tokens);
```
Comment (ondrejmirtes) [inline src/Parser/PhpDocParser.php]: Could we call the already existing `parseTemplateTagValue`?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Parser/PhpDocParser.php (hunk 4 above)
Comment (tyabu12) [reply src/Parser/PhpDocParser.php]: @ondrejmirtes 
I fixed it.
`parseTemplateTagValue` also parse `description` text, so it was necessary to add the flag parameter  `$parseDescription`.

**Comment by ondrejmirtes** [comment]
This is perfect, thank you very much :) I hope you follow-up with an implementation in phpstan-src :)

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
I see that this was added in phpdoc-parser over a year ago, but you never followed up with implementing this in phpstan-src. Can you please do? Thanks.

**Pull request merged**